### PR TITLE
Update number visual element with stepper button & shift key

### LINKED ISSUE #1703: [🐛 BUG] Number visual element does not update when using arrow to change value
### What went wrong? 🤔

The number visual element is not registering the update when using arrow up and down button to change value. It does update when inputing the value manually.

### Expected Behavior

The number arrow should change the value

### Steps to Reproduce Issue

```
from taipy.gui import Gui

a = 10

 def on_change(state, key, value, module):
     print(state, key, value, module)

page = """
<|{a}|>
<|{a}|number|>
"""
gui = Gui(page=page)

gui.run(run_browser=False)
```

### Solution Proposed

Should be a frontend issue only

### Screenshots

DESCRIPTION

### Runtime Environment

_No response_

### Browsers

_No response_

### OS

_No response_

### Version of Taipy

_No response_

### Additional Context

_No response_

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Create related issue in taipy-doc for documentation and Release Notes.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by namnguyen20999**
resolves https://github.com/Avaiga/taipy/issues/1703

```
from taipy.gui import Gui

x = 20

...
md = """
<|{x}|>
<|{x}|number|on_change=on_input_change|>
"""

...


def on_input_change(state):
    print(f"'x' was changed to: {state.x}")


...

Gui(page=md).run(use_reloader=True)
```

### CODE REVIEW

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 1]
```diff
@@ -112,6 +112,7 @@ const Input = (props: TaipyInputProps) => {
                         val = max;
                     }
                     setValue(val.toString());
+                    dispatch(createSendUpdateAction(updateVarName, val.toString(), module, onChange, propagate));
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx]: why val.tostring() ? if val is a number, all good ...

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 2]
```diff
@@ -171,24 +175,18 @@ const Input = (props: TaipyInputProps) => {
...
-                );
-                if (min !== undefined && Number(newValue) < min) {
-                    return min.toString();
-                }
-                if (max !== undefined && Number(newValue) > max) {
-                    return max.toString();
-                }
-                return newValue;
-            });
+            const newValue = calculateNewValue(value, step || 1, stepMultiplier || 10, event.shiftKey, increment);
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx]: you can dispatch from the setValue inner function
That allows to *not* depend on the value ie lee re-render ...

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 3]
```diff
@@ -121,8 +122,10 @@ const Input = (props: TaipyInputProps) => {
                         val = min;
                     }
                     setValue(val.toString());
+                    dispatch(createSendUpdateAction(updateVarName, val.toString(), module, onChange, propagate));
                     evt.preventDefault();
                 }
+                onAction && dispatch(createSendActionNameAction(id, module, onAction, evt.key, updateVarName, value));
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx]: I don't think `value` is the right thing here ?

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
dependency on value can be a problem

File: frontend/taipy-gui/src/components/Taipy/Input.tsx (hunk 2 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx]: I don't think it's a good idea, I tried dispatching it from the the setValue, it works however React gave me a warning on bad setState while rendering the component

Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx]: Only need a timeout

File: frontend/taipy-gui/src/components/Taipy/Input.tsx (hunk 1 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx]: updated

File: frontend/taipy-gui/src/components/Taipy/Input.tsx (hunk 2 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx]: timeout added

File: frontend/taipy-gui/src/components/Taipy/Input.tsx (hunk 3 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx]: I've changed it to the correct val

**Review by FabienLelaquais — DISMISSED** [review]
👍

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 4]
```diff
@@ -179,16 +204,22 @@ const Input = (props: TaipyInputProps) => {
                     event.shiftKey,
                     increment
                 );
+
                 if (min !== undefined && Number(newValue) < min) {
+                    updateValueWithDelay(min);
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx:209]: I guess you would still have the message "warning on bad setState while rendering the component" if changeDelay == 0 ?

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]

File: frontend/taipy-gui/src/components/Taipy/Input.tsx (hunk 4 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx:209]: it only show once then never show again so I think it would be fine

Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx:209]: If the user do not set the change_delay = 0 then it will use the 300ms default delay

**Review by FabienLelaquais — DISMISSED** [review]
👍

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 5]
```diff
@@ -80,14 +80,47 @@ const Input = (props: TaipyInputProps) => {
...
+        [props.width]
+    );
+
+    const updateValueWithDelay = useCallback(
+        (value: number | string) => {
+            if (changeDelay === -1) {
+                return;
+            }
+            if (changeDelay === 0) {
+                delayCall.current = window.setTimeout(() => {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx]: you don't need to set delayCall if changeDelay == 0

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 6]
```diff
@@ -80,14 +80,47 @@ const Input = (props: TaipyInputProps) => {
...
+        [changeDelay, dispatch, updateVarName, module, onChange, propagate]
+    );
 
     const handleInput = useCallback(
         (e: React.ChangeEvent<HTMLInputElement>) => {
             const val = e.target.value;
             setValue(val);
             if (changeDelay === 0) {
-                dispatch(createSendUpdateAction(updateVarName, val, module, onChange, propagate));
+                delayCall.current = window.setTimeout(() => {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx]: idem
and I think you don't need the setTimeout if handleInput is only called by user interaction (ie not in a setState)

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 7]
```diff
@@ -114,6 +147,8 @@ const Input = (props: TaipyInputProps) => {
                         val = max;
                     }
                     setValue(val.toString());
+                    updateValueWithDelay(val);
+                    onAction && dispatch(createSendActionNameAction(id, module, onAction, evt.key, updateVarName, val));
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx]: And I don't think we want to call the onAction when the user presses Arrow up or down ?

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 8]
```diff
@@ -123,6 +158,8 @@ const Input = (props: TaipyInputProps) => {
                         val = min;
                     }
                     setValue(val.toString());
+                    updateValueWithDelay(val);
+                    onAction && dispatch(createSendActionNameAction(id, module, onAction, evt.key, updateVarName, val));
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx]: idem

File: frontend/taipy-gui/src/components/Taipy/Input.tsx (hunk 4 above)
Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx:209]: one warning is not good enough ;-)

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
Almost there

File: frontend/taipy-gui/src/components/Taipy/Input.tsx (hunk 5 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx]: delayCall removed

**Review by FredLL-Avaiga — APPROVED** [review]

**Pull request merged**